From c6eb916f53f9fd3b4d77f8090b20a5788d177925 Mon Sep 17 00:00:00 2001
From: Su Dongming <sudongming1@huawei.com>
Date: Tue, 25 Apr 2023 17:11:24 +0800
Subject: [PATCH 0890/1646] [Huawei] drivers: hooks: add mmc driver vendor
 hooks

Offering: HULK
Hi171x_chip_solution inclusion
category: other
bugzilla: NA

----------------------------------------

Allow vendor modules to attach mmc driver hooks.

Signed-off-by: Su Dongming <sudongming1@huawei.com>
---
 drivers/hooks/Kconfig        |  9 +++++++++
 drivers/hooks/vendor_hooks.c |  5 +++++
 drivers/mmc/core/mmc.c       |  4 ++++
 include/trace/hooks/mmc.h    | 39 ++++++++++++++++++++++++++++++++++++
 4 files changed, 57 insertions(+)
 create mode 100644 include/trace/hooks/mmc.h

diff --git a/drivers/hooks/Kconfig b/drivers/hooks/Kconfig
index bd55ae4aa5c0..1357684a8171 100644
--- a/drivers/hooks/Kconfig
+++ b/drivers/hooks/Kconfig
@@ -77,4 +77,13 @@ config VENDOR_INET_HOOKS
             Allow vendor modules to attach inet hooks defined via
             DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
 
+config VENDOR_MMC_HOOKS
+        bool "MMC driver Vendor Hooks"
+        depends on VENDOR_HOOKS && MMC
+        default n
+        help
+            Enable mmc driver vendor hooks
+            Allow vendor modules to attach mmc driver hooks defined via
+            DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
+
 endmenu
diff --git a/drivers/hooks/vendor_hooks.c b/drivers/hooks/vendor_hooks.c
index d4251038740d..8eec47987aaa 100644
--- a/drivers/hooks/vendor_hooks.c
+++ b/drivers/hooks/vendor_hooks.c
@@ -14,6 +14,7 @@
 #include <trace/hooks/igb.h>
 #include <trace/hooks/i40e.h>
 #include <trace/hooks/inet.h>
+#include <trace/hooks/mmc.h>
 
 /*
  * Export tracepoints that act as a bare tracehook (ie: have no trace event
@@ -56,3 +57,7 @@ EXPORT_HOOK_SYMBOL_GPL(vendor_i40e_netdev_stats);
 #ifdef CONFIG_VENDOR_INET_HOOKS
 EXPORT_HOOK_SYMBOL_GPL(vendor_ip_finish_output);
 #endif
+
+#ifdef CONFIG_VENDOR_MMC_HOOKS
+EXPORT_HOOK_SYMBOL_GPL(vendor_mmc_init_card);
+#endif
diff --git a/drivers/mmc/core/mmc.c b/drivers/mmc/core/mmc.c
index 87807ef010a9..cbdf4faf3eae 100644
--- a/drivers/mmc/core/mmc.c
+++ b/drivers/mmc/core/mmc.c
@@ -17,6 +17,8 @@
 #include <linux/mmc/card.h>
 #include <linux/mmc/mmc.h>
 
+#include <trace/hooks/mmc.h>
+
 #include "core.h"
 #include "card.h"
 #include "host.h"
@@ -1837,6 +1839,8 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 	 * sudden power failure tests. Let's extend the timeout to a minimum of
 	 * DEFAULT_CACHE_EN_TIMEOUT_MS and do it for all cards.
 	 */
+	trace_vendor_mmc_init_card(card);
+
 	if (card->ext_csd.cache_size > 0) {
 		unsigned int timeout_ms = MIN_CACHE_EN_TIMEOUT_MS;
 
diff --git a/include/trace/hooks/mmc.h b/include/trace/hooks/mmc.h
new file mode 100644
index 000000000000..81de3395a7ca
--- /dev/null
+++ b/include/trace/hooks/mmc.h
@@ -0,0 +1,39 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2023-2023. All rights reserved.
+ * Description: Mmc driver Vendor Hooks
+ */
+
+#ifdef CONFIG_VENDOR_MMC_HOOKS
+
+#undef TRACE_SYSTEM
+#define TRACE_SYSTEM mmc
+
+#define TRACE_INCLUDE_PATH trace/hooks
+#if !defined(_TRACE_HOOK_MMC_HOOKS_H) || defined(TRACE_HEADER_MULTI_READ)
+#define _TRACE_HOOK_MMC_HOOKS_H
+#include <linux/tracepoint.h>
+#include <trace/hooks/vendor_hooks.h>
+
+struct mmc_card;
+
+DECLARE_HOOK(vendor_mmc_init_card,
+	TP_PROTO(struct mmc_card *card),
+	TP_ARGS(card));
+
+#endif
+/* This part must be outside protection */
+
+#include <trace/define_trace.h>
+
+#else
+
+#include <linux/tracepoint.h>
+
+struct mmc_card;
+DECLARE_EVENT_NOP(vendor_mmc_init_card,
+	TP_PROTO(struct mmc_card *card),
+	TP_ARGS(card));
+
+#endif /* CONFIG_VENDOR_MMC_HOOKS */
+
-- 
2.17.1

